tokio/src/fs/file.rs

```diff
@@ -756,3 +760,78 @@ impl Inner {
         }
     }
 }
+
+#[cfg(all(target_os = "linux", not(test)))]
+mod read_nowait {
+    use crate::io::ReadBuf;
+    use libc::{c_int, c_void, iovec, off_t, preadv2};
+    use std::{
+        os::unix::prelude::AsRawFd,
+        sync::atomic::{AtomicBool, Ordering},
+    };
+
+    static NONBLOCKING_READ_SUPPORTED: AtomicBool = AtomicBool::new(true);
+
+    pub(crate) fn try_nonblocking_read(
+        file: &crate::fs::sys::File,
+        dst: &mut ReadBuf<'_>,
+    ) -> Option<std::io::Result<()>> {
+        if !NONBLOCKING_READ_SUPPORTED.load(Ordering::Relaxed) {
+            return None;
+        }
+        let out = preadv2_safe(file, dst, -1, libc::RWF_NOWAIT);
+        if let Err(err) = &out {
+            match err.raw_os_error() {
+                Some(libc::ENOSYS) => {
+                    NONBLOCKING_READ_SUPPORTED.store(false, Ordering::Relaxed);
+                    return None;
+                }
```

Comment:
Have you tested this codepath on a machine that doesn't support the operation? All my linux machines are too recent for the purpose, but it would be good to test it.

Reply:
I just tried it on Ubuntu 14.04 (Linux 4.9.0, glibc 2.19) in a VM and I'm getting "undefined reference to `preadv2`" linker errors. This will be due to old glibc rather than lack of support in the kernel.

So I've got a bit more work to do here. I'll add a TODO item.  Maybe I need to add a fallback to libc, or maybe there's some way of doing compile time or runtime feature probing.

Reply:
Ah, interesting. I don't know how you would handle that. Maybe you can augment our `build.rs` script to do it? If there's a way to support it without causing breakage when run on a machine with an older libc than the one it was compiled on, that would be best, but I don't know if that is possible.

Reply:
Hmm, the libc crate seems to only support glibc 2.26+: https://github.com/rust-lang/libc/issues/1412 so maybe the test was invalid. I couldn't find any policy statement for tokio regarding old glibc support.  The closest I found was in this [mio announcement](https://tokio.rs/blog/2019-12-mio-v0.7-alpha.1):

>Support for the following OSes has been dropped:
>
> * Linux below version 2.6.27 (and glibc 2.9) ...

Does tokio have any official statement of what GLibc versions it should compile/run against?

Reply:
No, we don't declare the glibc version anywhere.

Reply:
One option would be to use `dlsym(RTLD_DEFAULT, "preadv2")` to get the symbol at runtime. I guess it would look something like:

```rust
    use libc::{c_int, iovec, off_t, ssize_t};
    type PReadV2 = fn(c_int, *const iovec, c_int, off_t, c_int) -> ssize_t;
    static PREADV2: AtomicPtr<PReadV2> = AtomicPtr<PReadV2>::new(0);
    let preadv2 = PREADV2.load(Ordering::Relaxed);
    if !preadv2 {
        preadv2 = libc::dlsym(libc::RTLD_DEFAULT, "preadv2");
        if preadv2 {
            PREADV2.store(preadv2, Ordering::Relaxed);
        } else {
            return Err(io::Error::from_os_error(libc::ENOSYS));
        }
    }
```

Although it's quite a lot of complexity for this edge case.  Using once_cell doesn't save much, but would have higher overheads:

```rust
    use libc::{c_int, iovec, off_t, ssize_t};
    type PReadV2 = fn(c_int, *const iovec, c_int, off_t, c_int) -> ssize_t;
    static PREADV2: OnceCell<PReadV2> = OnceCell<PReadV2>::new();
    let preadv2 = PREADV2.get_or_init(|| libc::dlsym(libc::RTLD_DEFAULT, "preadv2"));
    if !preadv2 {
        return Err(io::Error::from_os_error(libc::ENOSYS));
    }
```

Maybe there's some sort of linker trick we could play such that preadv2 takes its implementation from us if it's not found externally, like a (very) weak symbol?

Reply:
Looks like linkage is unstable: https://github.com/rust-lang/rust/issues/29603 :(

I need some advice here. I'm not sure what to do. It all depends on what's acceptable to you - the tokio maintainers.

Reply:
Using `dlsym` with an atomic is acceptable to me if it works.

Reply:
Note that `AtomicPtr<PReadV2>` is not the correct type here. That corresponds to a pointer to a function pointer. The easiest is probably to just use `AtomicUsize` along with a cast. Then we can also use 1 as a value that means "we know that it is not supported". See [this page](https://rust-lang.github.io/unsafe-code-guidelines/layout/function-pointers.html) for info on the representation of function pointers.

Reply:
Well, I've done it, but I don't like it. There's a bunch of `transmute`s :(. I wonder what will happen if one tries to statically link now seeing as we're calling `dlsym`?

There has to be a better way. I'll investigate `build.rs` for compile time detection. The other alternative would be to call the `SYS_preadv2` syscall directly via `libc::syscall`. The advantage being that it would work with all libcs on Linux including musl, bionic, etc as it basically won't be using libc. The downside would be that it might require architecture specific code.

Reply:
Hmm, this is interesting: https://doc.rust-lang.org/stable/src/std/sys/unix/weak.rs.html

Reply:
impl using `dlsym`: 5bd8ea4e722a7cfddbe2a0c228c7a84de5a02691
impl using `syscall`: 6ff49323c5cc6a8529a8e83815b02c1ec526d1e2

Reply:
It seems like libc has already taken care of most of the architecture specific stuff by defining the `libc::SYS_preadv2` constant. Is there anything else we would need to handle?

Also, I see that you suggested doing the syscall solution in the libc crate in https://github.com/rust-lang/libc/pull/2066#issuecomment-777567493.